### PR TITLE
Adding PHP Generator Support

### PR DESCRIPTION
**Pull request by daarond**

### CODE REVIEW

File: generators/php.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
+/**
+ * @license
+ * Visual Blocks Language
+ *
+ * Copyright 2012 Google Inc.
```
Comment (NeilFraser) [inline generators/php.js]: New files should be dated with the current year 2015.  [Change here and all other new files.]

You could also change Google Inc to be yourself or your company, but that would mean our lawyers would need to contact you with stuff to sign.  Thanks for leaving this as Google, with yourself as the author below.  [No change needed.]

File: generators/php.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+/**
+ * List of illegal variable names.
+ * This is not intended to be a security feature.  Blockly is 100% client-side,
+ * so bypassing this list is trivial.  This is intended to prevent users from
+ * accidentally clobbering a built-in object or function.
+ * @private
+ */
+Blockly.PHP.addReservedWords(
+    'Blockly,' +  // In case JS is evaled in the current window.
```
Comment (NeilFraser) [inline generators/php.js]: 'Blockly' reserved word only needed in JavaScript.  [Remove line.]

File: generators/php.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        // http://php.net/manual/en/reserved.constants.php
+    'PHP_VERSION,PHP_MAJOR_VERSION,PHP_MINOR_VERSION,PHP_RELEASE_VERSION,PHP_VERSION_ID,PHP_EXTRA_VERSION,PHP_ZTS,PHP_DEBUG,PHP_MAXPATHLEN,PHP_OS,PHP_SAPI,PHP_EOL,PHP_INT_MAX,PHP_INT_SIZE,DEFAULT_INCLUDE_PATH,PEAR_INSTALL_DIR,PEAR_EXTENSION_DIR,PHP_EXTENSION_DIR,PHP_PREFIX,PHP_BINDIR,PHP_BINARY,PHP_MANDIR,PHP_LIBDIR,PHP_DATADIR,PHP_SYSCONFDIR,PHP_LOCALSTATEDIR,PHP_CONFIG_FILE_PATH,PHP_CONFIG_FILE_SCAN_DIR,PHP_SHLIB_SUFFIX,E_ERROR,E_WARNING,E_PARSE,E_NOTICE,E_CORE_ERROR,E_CORE_WARNING,E_COMPILE_ERROR,E_COMPILE_WARNING,E_USER_ERROR,E_USER_WARNING,E_USER_NOTICE,E_DEPRECATED,E_USER_DEPRECATED,E_ALL,E_STRICT,__COMPILER_HALT_OFFSET__,TRUE,FALSE,NULL,__CLASS__,__DIR__,__FILE__,__FUNCTION__,__LINE__,__METHOD__,__NAMESPACE__,__TRAIT__');
+
+/**
+ * Order of operation ENUMs.
+ * https://developer.mozilla.org/en/PHP/Reference/Operators/Operator_Precedence
```
Comment (NeilFraser) [inline generators/php.js]: 404 at mozilla.org

File: generators/php.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+Blockly.PHP.ORDER_NEW = 1;            // new
+Blockly.PHP.ORDER_MEMBER = 2;         // ()
+Blockly.PHP.ORDER_FUNCTION_CALL = 2;  // ()
+Blockly.PHP.ORDER_INCREMENT = 3;      // ++
+Blockly.PHP.ORDER_DECREMENT = 3;      // --
+Blockly.PHP.ORDER_LOGICAL_NOT = 4;    // !
+Blockly.PHP.ORDER_BITWISE_NOT = 4;    // ~
+Blockly.PHP.ORDER_UNARY_PLUS = 4;     // +
+Blockly.PHP.ORDER_UNARY_NEGATION = 4; // -
+Blockly.PHP.ORDER_TYPEOF = 4;         // typeof
```
Comment (NeilFraser) [inline generators/php.js]: Do these orders even exist in PHP?  typeof is not in the reserved word list.

File: generators/php.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  var safeName = this.variableDB_.safeName_(name);
+  var i = '';
+  while ((type == Blockly.Procedures.NAME_TYPE && (safeName + i) in this.variableDB_.reservedDict_)
+  || this.variableDB_.dbReverse_[safeName + i]) {
+    // Collision with existing name.  Create a unique name.
+    i = i ? i + 1 : 2;
+  }
+  safeName += i;
+  this.variableDB_.dbReverse_[safeName] = true;
+  if (type === Blockly.Procedures.NAME_TYPE) return safeName;
```
Comment (NeilFraser) [inline generators/php.js]: Suggest reversing this:
if (type == Blockly.Variables.NAME_TYPE) {
  safeName =  '$' + safeName;
}
return safeName;

File: msg/json/en.json [hunk 6]
```diff
@@ -1,370 +1,9 @@
...
-	"PROCEDURES_CALLRETURN_TOOLTIP": "Run the user-defined function '%1' and use its output.",
-	"PROCEDURES_MUTATORCONTAINER_TITLE": "inputs",
-	"PROCEDURES_MUTATORCONTAINER_TOOLTIP": "Add, remove, or reorder inputs to this function.",
-	"PROCEDURES_MUTATORARG_TITLE": "input name:",
-	"PROCEDURES_MUTATORARG_TOOLTIP": "Add an input to the function.",
-	"PROCEDURES_HIGHLIGHT_DEF": "Highlight function definition",
-	"PROCEDURES_CREATE_DO": "Create '%1'",
-	"PROCEDURES_IFRETURN_TOOLTIP": "If a value is true, then return a second value.",
-	"PROCEDURES_IFRETURN_WARNING": "Warning: This block may be used only within a function definition."
+
```
Comment (NeilFraser) [inline msg/json/en.json]: Why delete all the messages?  Something bad happened in your compile?

File: msg/json/qqq.json [hunk 7]
```diff
@@ -1,371 +1,3 @@
...
-	"LISTS_GET_SUBLIST_TAIL": "block text - This appears in the rightmost position ('tail') of the sublist block, as described at [https://github.com/google/blockly/wiki/Lists#getting-a-sublist https://github.com/google/blockly/wiki/Lists#getting-a-sublist]. In English and most other languages, this is the empty string. [[File:Blockly-get-sublist.png]]",
-	"LISTS_GET_SUBLIST_TOOLTIP": "tooltip - See [https://github.com/google/blockly/wiki/Lists#getting-a-sublist https://github.com/google/blockly/wiki/Lists#getting-a-sublist] for more information. [[File:Blockly-get-sublist.png]]",
-	"LISTS_SPLIT_HELPURL": "url - Information describing splitting text into a list, or joining a list into text.",
-	"LISTS_SPLIT_LIST_FROM_TEXT": "dropdown - Indicates that text will be split up into a list (e.g. 'a-b-c' -> ['a', 'b', 'c']).",
-	"LISTS_SPLIT_TEXT_FROM_LIST": "dropdown - Indicates that a list will be joined together to form text (e.g. ['a', 'b', 'c'] -> 'a-b-c').",
-	"LISTS_SPLIT_WITH_DELIMITER": "block text - Prompts for a letter to be used as a separator when splitting or joining text.",
-	"LISTS_SPLIT_TOOLTIP_SPLIT": "tooltip - See [https://github.com/google/blockly/wiki/Lists#make-l
```
Comment (NeilFraser) [inline msg/json/qqq.json]: This file is somewhat useful...

**Comment by NeilFraser** [comment]
Can you also edit tests/generators to add a PHP test generator and verify that they all pass?
https://developers.google.com/blockly/hacking/unit-testing
Thanks.

File: generators/php.js (hunk 4 above)
Comment (daarond) [inline generators/php.js]: You're right, but I was just leaving them, even if they were unused.

File: msg/json/en.json (hunk 6 above)
Comment (daarond) [inline msg/json/en.json]: Yes, apparently my compile is wonky.

File: msg/json/qqq.json (hunk 7 above)
Comment (daarond) [inline msg/json/qqq.json]: I think this is a function of the wonky compile. I'll look around to see if I can get that fixed.

File: msg/json/en.json (hunk 6 above)
Comment (daarond) [inline msg/json/en.json]: It appears to be an issue with /msg/json/en.js having the trailing comma. Removing it made the compile happy.

**Comment by NeilFraser** [comment]
Hi daarond,
Do you want to reply to this guy:
https://groups.google.com/forum/#!topic/blockly/OdiWXX7CjBQ
It is your code, and he's willing to pay.  :)  Have fun!

**Comment by daarond** [comment]
Sure, I'll reply to him. I have a bit more debugging to do, then I will
have something for you to look at.

**Comment by daarond** [comment]
All tests pass and I believe I have fixed the outstanding items, Neil.
Please let me know if there are any issues.

**Pull request closed without merging**